cvmfs/swissknife_graft.cc

```diff
@@ -0,0 +1,249 @@
+/**
+ * This file is part of the CernVM File System.
+ *
+ * Process a set of input files and create appropriate graft files.
+ */
+
+#include <fcntl.h>
+#include <unistd.h>
+#include <sys/stat.h>
+
+#include <cstdio>
+#include <iostream>
+#include <vector>
+#include <sstream>
+
+#include "compression.h"
+#include "fs_traversal.h"
+#include "hash.h"
+
+#include "swissknife_graft.h"
+
+#define CHUNK_SIZE 16*1024
+
+int swissknife::CommandGraft::ChecksumFd(int fd, shash::Any &id, off_t &processed_size) {  // NOLINT
+  shash::ContextPtr hash_context(id.algorithm);
+  hash_context.buffer = alloca(hash_context.size);
+  shash::Init(hash_context);
+
+  z_stream strm;
+  zlib::CompressInit(&strm);
+  zlib::StreamStates retval;
+
+  std::vector<unsigned char> buf; buf.reserve(CHUNK_SIZE);
+  off_t cksum_size = 0;
+  bool eof;
+
+  do {
+    ssize_t nbytes = read(fd, &buf[0], CHUNK_SIZE);
+    if (nbytes < 0) {
+      if (errno == EINTR) {continue;}
+      zlib::CompressFini(&strm);
+      return nbytes;
+    }
+    cksum_size += nbytes;
+    eof = nbytes == 0;
+    retval = zlib::CompressZStream2Null(&buf[0], nbytes, eof, &strm,
+                                        &hash_context);
+    if (retval == zlib::kStreamDataError) {
+      zlib::CompressFini(&strm);
+      errno = EINVAL;
+      return -1;
+    }
+  } while (!eof);
+
+  zlib::CompressFini(&strm);
+  if (retval != zlib::kStreamEnd) {
+    errno = EINVAL;
+    return -1;
+  }
+  shash::Final(hash_context, &id);
+  processed_size = cksum_size;
+  return 0;
+}
+
+
+static void splitpath(const std::string &path,
```

Comment:
I'd like to have a single function which pulls out both parts of the path.  I'll move this to `util.cc`.